### PR TITLE
[KED-2982] Re-write Kedro Dropdown and Menu-option

### PR DESCRIPTION
**Pull request by rashidakanchwala**
## Description

This PR is a part of the Kedro-UI migration ticket. In this PR we migrate the dropdown and menu-options functionality from Kedro-ui to Kedro-viz

## Checklist

- [ ] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

**Review by studioswong — CHANGES_REQUESTED** [review]
Thanks for setting this up. 

On testing the new dropdown the 'expand more' arrow is current missing - please see below

Dropdown in this PR:
image

Dropdown in current demo:
image

Part of the main goal of this migration work is to refactor the components and bringing the code up to date with the current react setup in our codebase, which includes stripping away the react lifecycle and local state setup to use react hooks. As per my comments below, please refactor the component with the new react hooks setup.

Thank you!

File: src/components/dropdown/dropdown.js [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+   * {@link https://facebook.github.io/react/docs/react-component.html#componentwillupdate}
+   * @param {Object} New component props
+   */
+  componentDidUpdate(prevProps) {
+    if (this._childrenHaveChanged(prevProps)) {
+      this.setState({
+        selectedOption: this._findSelectedOption(prevProps),
+      });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (rashidakanchwala) [reply src/components/dropdown/dropdown.js:47]: Thanks, I have moved this PR to draft stage now, and I will work on changing to hooks before releasing it for review again.

**Comment by tynandebold** [comment]
@studioswong after looking at this for an hour just now, I'm of the opinion we shouldn't convert this to use hooks at this time. The amount of effort Rashida and I have put in to this is much too great a cost for what we gain, which really isn't anything other than a stylistic alignment of the codebase to use one React pattern. I'm confident we'll be able to live with two components still using the `class` style, the `App` component being the other one.

Please let us know what you think. Again, I strongly feel we should leave it as a class given the amount of work involved. We'll then be able to finish the migration epic, and maybe even the sprint. We can revisit this and the `App` conversion at a later date, as one ticket.

cc @yetudada @idanov

**Comment by studioswong** [comment]
> 

Thanks for looking into this @tynandebold - may I ask what is the main problem you and Rashida encountered in the refactoring to use hooks?

I understand this for now may seem to be a matter of code style, however this has implications down the line as we maintain this piece of work later on - do we still continue down the line of deviating this component from hooks? Do we need to at some point just bite the bullet to refactor this? Don't want to be a blocker, but if we are doing the migration we might as well do it now while we are at it. ( broken window analogy, if you will.)

Regardless, I will check out this branch and take a look 😃

**Comment by tynandebold** [comment]
> > 
> 
> Thanks for looking into this @tynandebold - may I ask what is the main problem you and Rashida encountered in the refactoring to use hooks?
> 
> I understand this for now may seem to be a matter of code style, however this has implications down the line as we maintain this piece of work later on - do we still continue down the line of deviating this component from hooks? Do we need to at some point just bite the bullet to refactor this? Don't want to be a blocker, but if we are doing the migration we might as well invest a bit more time to do it now while we are at it. ( broken window analogy, if you will.)
> 
> Regardless, I will check out this branch and take a look 😃
> 
> cc @yetudada @idanov

The main problem is time invested versus payoff. I think there are other things we can do with our time that will benefit the product and thus our users. I also don't remember both the migration of these components to Viz _and also updating the pattern of how they are written_ as part of the epic. I thought it was just the former, though I do agree it should be aligned.

And to answer your questions:

> do we still continue down the line of deviating this component from hooks?

No, we combine this rewrite and the one for the `App` component together in one ticket since it's now taking the time of three developers to do.

> Do we need to at some point just bite the bullet to refactor this?

Yes, as stated above.

**Comment by studioswong** [comment]
> The main problem is time invested versus payoff. I think there are other things we can do with our time that will benefit the product and thus our users. I also don't remember both the migration of these components to Viz _and also updating the pattern of how they are written_ as part of the epic. I thought it was just the former, though I do agree it should be aligned.
> 

If we recall from the beginning, I actually don't think the idea of doing this migration is a good idea exactly given the problems we face now ( the need to migrate / refactor and maintain legacy react code in our codebase). But I understand there are benefits down the line ( such as being able to have full control and modify this component without the need to publish new versions of the kedro-ui npm package), and it is exactly for this reason to be able to maintain and make changes to this code that we have to do the rewrite. We are just biting the bullet for bigger rewards later on.

> And to answer your questions:
> 
> > do we still continue down the line of deviating this component from hooks?
> 
> No, we combine this rewrite and the one for the `App` component together in one ticket since it's now taking the time of three developers to do.
> 

Excuse me if I don't exactly understand what you meant here, but last when I checked out the branch and began working on it, the component was in a weird state with half migrated hooks and needs to fix anyways, so might as well.

**Comment by studioswong** [comment]
on another note, @rashidakanchwala @tynandebold I am currently refactoring dropdown.js into hooks now, so no need to worry about this - I'll push the changes once it's done

**Comment by studioswong** [comment]
@rashidakanchwala I had pushed the latest changes in refactoring `dropdown.js` into react hooks, as well as updates to the test given the relevant changes. 

There is still the problem with the missing 'expand more' arrow on the right of the dropdown, which was a seperate known issue to fix - I'll let you take back over for that one :)

File: src/components/dropdown/dropdown.js [hunk 2]
```diff
@@ -288,11 +288,9 @@ const Dropdown = (props) => {
...
     setOpen(false);
 
-    this.setState({ open: false }, () => {
-      if (typeof onClosed === 'function') {
-        onClosed();
-      }
-    });
+    if (typeof onClosed === 'function') {
+      onClosed();
+    }
```
Comment (studioswong) [inline src/components/dropdown/dropdown.js:293]: This is a callback on setting the state that needs to be handled in a useEffect to ensure this is only executed after the state change (i.e the `setOpen` state change here). 

Regardless, on checking the code, this is actually already set up in the useEffects above, hence can be deleted.

**Review by studioswong — APPROVED** [review]
looking good and all works the same now.

**Pull request merged**